File: documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/ExternPolicy.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.documentapi.messagebus.protocol;

import com.yahoo.jrt.Supervisor;
import com.yahoo.jrt.Transport;
import com.yahoo.jrt.slobrok.api.Mirror;
import com.yahoo.jrt.slobrok.api.SlobrokList;
import com.yahoo.messagebus.ErrorCode;
import com.yahoo.messagebus.routing.Hop;
import com.yahoo.messagebus.routing.Route;
import com.yahoo.messagebus.routing.RoutingContext;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * This policy implements the necessary logic to communicate with an external Vespa application and resolve its list of
 * recipients using that other application's slobrok servers.
 *
 * @author Simon Thoresen Hult
 */
public class ExternPolicy implements DocumentProtocolRoutingPolicy {

    private Supervisor orb = null;
    private Mirror mirror = null;
    private String pattern = null;
    private String session = null;
    private final String error;
    private int offset = 0;
    private int generation = 0;
    private final List<Hop> recipients = new ArrayList<>();
    private final AtomicBoolean destroyed = new AtomicBoolean(false);

    /**
     * Constructs a new instance of this policy. The argument given is the connection spec to the slobrok to use for
     * resolving recipients, as well as the pattern to use when querying. This constructor does _not_ wait for the
     * mirror to become ready.
     *
     * @param arg The slobrok connection spec.
     */
    public ExternPolicy(String arg) {
        if (arg == null || arg.length() == 0) {
            error = "Expected parameter, got empty string.";
            return;
        }
        String[] args = arg.split(";", 2);
        if (args.length != 2 || args[0].length() == 0 || args[1].length() == 0) {
            error = "Expected parameter on the form '<spec>;<pattern>', got '" + arg + "'.";
            return;
        }
        int pos = args[1].lastIndexOf('/');
        if (pos < 0) {
            error = "Expected pattern on the form '<service>/<session>', got '" + args[1] + "'.";
            return;
        }
        SlobrokList slobroks = new SlobrokList();
        slobroks.setup(args[0].split(","));
        pattern = args[1];
        session = pattern.substring(pos);
        orb = new Supervisor(new Transport("externpolicy"));
        orb.setDropEmptyBuffers(true);
        mirror = new Mirror(orb, slobroks);
        error = null;
    }

    /**
     * This is a safety mechanism to allow the constructor to fail and signal that it can not be used.
     *
     * @return The error string, or null if no error.
     */
    public String getError() {
        return error;
    }

    /**
     * Returns the slobrok mirror used by this policy to resolve external recipients.
     *
     * @return The external mirror.
     */
    public Mirror getMirror() {
        return mirror;
    }

    /**
     * Returns the appropriate recipient hop. This method provides synchronized access to the internal mirror.
     *
     * @return The recipient hop to use.
     */
    private synchronized Hop getRecipient() {
        update();
        if (recipients.isEmpty()) {
            return null;
        }
        int offset = ++this.offset & Integer.MAX_VALUE; // mask signed bit because of modulo
        return new Hop(recipients.get(offset % recipients.size()));
    }

    /**
     * Updates the list of matching recipients by querying the extern slobrok.
     */
    private void update() {
        int upd = mirror.updates();
        if (generation != upd) {
            generation = upd;
            recipients.clear();
            List<Mirror.Entry> arr = mirror.lookup(pattern);
            for (Mirror.Entry entry : arr) {
                recipients.add(Hop.parse(entry.getSpecString() + session));
            }
        }
    }

    @Override
    public void select(RoutingContext ctx) {
        if (error != null) {
            ctx.setError(DocumentProtocol.ERROR_POLICY_FAILURE, error);
        } else if (mirror.ready()) {
            Hop hop = getRecipient();
            if (hop != null) {
                Route route = new Route(ctx.getRoute());
                route.setHop(0, hop);
                ctx.addChild(route);
            } else {
                ctx.setError(ErrorCode.NO_ADDRESS_FOR_SERVICE,
                             "Could not resolve any recipients from '" + pattern + "'.");
            }
        } else {
            ctx.setError(ErrorCode.APP_TRANSIENT_ERROR, "Extern slobrok not ready.");
        }
    }

    @Override
    public void merge(RoutingContext ctx) {
        DocumentProtocol.merge(ctx);
    }

    @Override
    public void destroy() {
        if (destroyed.getAndSet(true)) throw new RuntimeException("Already destroyed");
        mirror.shutdown();
        orb.transport().shutdown().join();
    }
}

File: documentapi/src/vespa/documentapi/messagebus/policies/externpolicy.cpp
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
#include "externpolicy.h"
#include "mirror_with_all.h"
#include <vespa/vespalib/text/stringtokenizer.h>
#include <vespa/documentapi/messagebus/documentprotocol.h>
#include <vespa/vespalib/util/stringfmt.h>
#include <vespa/slobrok/sbmirror.h>
#include <vespa/fnet/transport.h>
#include <vespa/fnet/frt/supervisor.h>

#include <vespa/log/log.h>
LOG_SETUP(".externpolicy");

using slobrok::api::IMirrorAPI;
using slobrok::api::MirrorAPI;

namespace documentapi {

ExternPolicy::ExternPolicy(const string &param) :
    _lock(),
    _mirrorWithAll(),
    _pattern(),
    _session(),
    _error("Not initialized."),
    _offset(0),
    _gen(0),
    _recipients()
{
    // Parse connection spec.
    if (param.empty()) {
        _error = "Expected parameter, got empty string.";
        return;
    }
    size_t pos = param.find(';');
    if (pos == string::npos || pos == 0 || pos == param.size() - 1) {
        _error = vespalib::make_string("Expected parameter on the form '<spec>;<pattern>', got '%s'.", param.c_str());
        return;
    }

    // Activate supervisor and register mirror.
    MirrorAPI::StringList spec;
    vespalib::string lst = param.substr(0, pos);
    vespalib::StringTokenizer slobrokList(lst, ",");
    for (uint32_t j = 0; j < slobrokList.size(); j++) {
        spec.push_back(slobrokList[j]);
    }

    if (spec.empty()) {
        _error = vespalib::make_string("Extern policy needs at least one slobrok: Slobrok list '%s' resolved to no slobroks", lst.c_str());
        return;
    }

    slobrok::ConfiguratorFactory config(spec);
    _mirrorWithAll = std::make_unique<MirrorAndStuff>(config);
     LOG(debug, "Connecting to extern slobrok mirror '%s'..", lst.c_str());


    // Parse query pattern.
    _pattern = param.substr(pos + 1);
    pos = _pattern.find_last_of('/');
    if (pos == string::npos) {
        _error = vespalib::make_string("Expected pattern on the form '<service>/<session>', got '%s'.", _pattern.c_str());
        return;
    }
    _session = _pattern.substr(pos);

    // All ok.
    _error.clear();
}

ExternPolicy::~ExternPolicy() = default;

const IMirrorAPI *
ExternPolicy::getMirror() const {
    return _mirrorWithAll ? _mirrorWithAll->mirror() : nullptr;
}

void
ExternPolicy::select(mbus::RoutingContext &ctx)
{
    if (!_error.empty()) {
        ctx.setError(DocumentProtocol::ERROR_POLICY_FAILURE, _error);
    } else if (_mirrorWithAll->mirror()->ready()) {
        mbus::Hop hop = getRecipient();
        if (hop.hasDirectives()) {
            mbus::Route route = ctx.getRoute();
            route.setHop(0, hop);
            ctx.addChild(route);
        } else {
            ctx.setError(mbus::ErrorCode::NO_ADDRESS_FOR_SERVICE,
                         vespalib::make_string("Could not resolve any recipients from '%s'.", _pattern.c_str()));
        }
    } else {
        ctx.setError(mbus::ErrorCode::APP_TRANSIENT_ERROR, "Extern slobrok not ready.");
    }

}

void
ExternPolicy::merge(mbus::RoutingContext &ctx)
{
    DocumentProtocol::merge(ctx);
}

mbus::Hop
ExternPolicy::getRecipient()
{
    std::lock_guard guard(_lock);
    update();
    if (_recipients.empty()) {
        return mbus::Hop();
    }
    return _recipients[++_offset % _recipients.size()];
}

void
ExternPolicy::update()
{
    uint32_t upd = _mirrorWithAll->mirror()->updates();
    if (_gen != upd) {
        _gen = upd;
        _recipients.clear();

        IMirrorAPI::SpecList entries = _mirrorWithAll->mirror()->lookup(_pattern);
        if (!entries.empty()) {
            for (const auto & spec : entries) {
                _recipients.push_back(mbus::Hop::parse(spec.second + _session));
            }
        }
    }
}

}
